template-compose/data/src/main/java/co/nimblehq/template/compose/data/repository/UserPreferencesRepositoryImpl.kt

```diff
@@ -0,0 +1,54 @@
+package co.nimblehq.template.compose.data.repository
+
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.*
+import co.nimblehq.template.compose.domain.model.UserPreferences
+import co.nimblehq.template.compose.domain.repository.UserPreferencesRepository
+import kotlinx.coroutines.flow.*
+import java.io.IOException
+import javax.inject.Inject
+
+class UserPreferencesRepositoryImpl @Inject constructor(private val dataStore: DataStore<Preferences>) :
+    UserPreferencesRepository {
+
+    private val TAG: String = "UserPreferencesRepo"
+
+    private object PreferencesKeys {
+        val USERNAME = stringPreferencesKey("username")
+        val IS_ACTIVE = booleanPreferencesKey("is_active")
+    }
+
+    override fun getUsernamePreferences(): Flow<UserPreferences> = dataStore.data
+        .catch { exception ->
+            // dataStore.data throws an IOException when an error is encountered when reading data
+            if (exception is IOException) {
+                Log.e(TAG, "Error reading preferences.", exception)
+                emit(emptyPreferences())
```

Comment:
@lydiasama I don't have a similar with `DataStore`, So, May I know why we prefer to emit `emptyPreferences()` instead of `throw exception` when we got `IOException` 🙏

Reply:
@Wadeewee Because this is a template, so I don't think we need to handle the exception. 🙏🏻

Reply:
@Wadeewee For more information
![Screenshot 2566-01-03 at 2 42 18 PM](https://user-images.githubusercontent.com/12026942/210316575-f78c9e68-ef25-4120-90f3-02d71f03bdbd.png)

Ref: https://developer.android.com/codelabs/android-preferences-datastore#5